mat64/matrix_test.go

```diff
@@ -58,15 +59,57 @@ func eye() *Dense {
 }
 
 func TestCol(t *testing.T) {
-	f := func(a Matrix) interface{} {
-		_, c := a.Dims()
+	for id, af := range [][][]float64{
+		{
+			{1, 2, 3},
+			{4, 5, 6},
+			{7, 8, 9},
+		},
+		{
+			{1, 2, 3},
+			{4, 5, 6},
+			{7, 8, 9},
+			{10, 11, 12},
+		},
+		{
+			{1, 2, 3, 4},
+			{5, 6, 7, 8},
+			{9, 10, 11, 12},
+		},
+	} {
+		a := NewDense(flatten(af))
+		col := make([]float64, a.mat.Rows)
+		for j := range af[0] {
+			for i := range col {
+				col[i] = float64(i*a.mat.Cols + j + 1)
+			}
+
+			if got := Col(nil, j, a); !reflect.DeepEqual(got, col) {
+				t.Errorf("test %d: unexpected values returned for dense col %d: got: %v want: %v",
+					id, j, got, col)
+			}
+
+			got := make([]float64, a.mat.Rows)
+			if Col(got, j, a); !reflect.DeepEqual(got, col) {
+				t.Errorf("test %d: unexpected values filled for dense col %d: got: %v want: %v",
+					id, j, got, col)
+			}
+		}
+	}
+
+	denseComparison := func(a *Dense) interface{} {
+		r, c := a.Dims()
 		ans := make([][]float64, c)
 		for j := range ans {
-			ans[j] = Col(nil, j, a)
+			ans[j] = make([]float64, r)
+			for i := range ans[j] {
+				ans[j][i] = a.At(i, j)
```

Comment:
It seemed to me that the comparison should not use the function that is being tested.